File: Baes.cpp
#define _USE_MATH_DEFINES
#include "Baes.h"
#include "include/libstemmer.h"
#include <iostream>
#include <fstream>
#include <map>
#include <ctime>
#include <set>
#include <cmath>

#define M_E 2.71828182845904523536
#include "Util.h"

/// Read N texts from positive and negative files for training
void Baes::GetTexts(int count)
{
	int i = 0;
	wstring line;
	wifstream positive_file(positive_path, ios::in);
	while (getline(positive_file, line) && i < count)
	{
		try
		{
			auto text = Split(line, L';');
			if (text.size() > 3)
				positive_text.push_back(text[3].substr(1, text[3].length() - 1));
		}
		catch (exception& ex)
		{
			// Possible missed line
			cerr << ex.what() << endl;
		}
		i++;
	}
	positive_file.close();

	i = 0;
	wifstream negative_file(negative_path, ios::in);
	while (getline(negative_file, line) && i < count)
	{
		try
		{
			auto text = Split(line, L';');
			if (text.size() > 3)
				negative_text.push_back(text[3].substr(1, text[3].length() - 1));
		}
		catch (exception& ex)
		{
			// Possible missed line
			cerr << ex.what() << endl;
		} 
		i++;
	}
	negative_file.close();
}

/// Split words in text
vector<wstring> Baes::SplitWords(const vector<wstring>& text)
{
	vector<wstring> words;
	for (auto &t : text)
	{
		auto splited = Split(t, L' ');
		words.insert(words.end(), splited.begin(), splited.end());
	}

	return words;
}

/// Split words by delimiter
vector<wstring> Baes::Split(const wstring &str, wchar_t delimiter)
{
	vector<wstring> internal;
	wstringstream ss(str);
	wstring tok;

	while (getline(ss, tok, delimiter)) {
		internal.push_back(tok);
	}

	return internal;
}

/// Get unique words count in texts
size_t Baes::GetUniqueCount(vector<wstring>& negative_words, vector<wstring>& positive_words)
{
	set<wstring> unique;
	for (auto &w : negative_words)
		unique.insert(w);

	for (auto &w : positive_words)
		unique.insert(w);

	return unique.size();
}

/// Russian Porter's stemming (snowball) 
void Baes::Stemming(vector<wstring>& words)
{
	auto stemmer = sb_stemmer_new("russian", "UTF_8"); 
	if (stemmer == nullptr)
	{
		wcerr << L"Can't allocate stemmer" << endl;
		return;
	}
	
	for (auto &w : words)
	{
		auto stem = sb_stemmer_stem(stemmer, reinterpret_cast<const sb_symbol *>(w.c_str()), w.size()*2);
		string str = reinterpret_cast<const char *>(stem);
		//w = Util::ToWide(str);
	}

	sb_stemmer_delete(stemmer);
}

Baes::Baes(const wstring& negative_path, const wstring& positive_path) :positive_path(positive_path), negative_path(negative_path), _trained(false)
{
}

/// Training model from negative and positive files
void Baes::Train(int count = -1)
{
	// Reading files
	wcout << L"Training started..." << endl;
	float start_time = clock();
	GetTexts(count);
	wcout << L"Reading texts: " << double(clock() - start_time) / CLOCKS_PER_SEC << " sec." << endl;

	// Splited words
	negative_words = SplitWords(negative_text);
	positive_words = SplitWords(positive_text);
	
	// Stemming
	start_time = clock();
	Stemming(negative_words);
	Stemming(positive_words);

	wcout << L"Stemming words: " << double(clock() - start_time) / CLOCKS_PER_SEC << " sec." << endl;

	// Filter for stop-words, symbols, links
	// Filter(negative_words);
	// Filter(positive_words);

	// Количество текстов в обучающей выборке в классе negative
	Dn = negative_text.size();
	// Количество текстов в обучающей выборке в классе positive
	Dp = positive_text.size();
	// Общее число текстов в обучающей выборке
	D = Dn + Dp;
	// Количество уникальных слов во всех текстах
	V = GetUniqueCount(negative_words, positive_words);
	// Суммарное количество слов в текстах класса negative
	Ln = negative_words.size();
	// Суммарное количество слов в текстах класса positive
	Lp = positive_words.size();

	_trained = true;

	wcout << L"Training end." << endl;

#ifdef DEBUG
	wofstream of(L"output", ios::out);
	for (auto &t : negative_text)
	{
		of << t << endl;
	}
#endif
}

/// Saving model
void Baes::Save(wstring& file_name)
{
	// TODO Save model
	wcout << L"Saving model..." << endl;
	wofstream of(file_name,  ios::binary);
	of << negative_text.size() << endl;
	for (auto &t: negative_text)
		of << t << endl;
	
	of << positive_text.size() << endl;
	for (auto &t : positive_text)
		of << t << endl;

	of << negative_words.size() << endl;
	for (auto &t : negative_words)
		of << t << endl;

	of << positive_words.size() << endl;
	for (auto &t : positive_words)
		of << t << endl;
	of << Dn << endl << Dp << endl << D << endl << V << endl << Ln << endl << Lp << endl;

	of.close();
	wcout << L"Model saved to: " << file_name << endl;
}

/// Loading model from file
bool Baes::Load(wstring& file_name)
{
	_trained = false;

	wcout << L"Model loading..." << endl;
	
	wifstream file(file_name, ios::in);
	wstring line{};

	try
	{
		// negative text
		getline(file, line);
		auto nt_count = stoi(line);
		negative_text.clear();
		for (auto i = 0; i < nt_count; ++i)
		{
			getline(file, line);
			negative_text.push_back(line);
		}

		// positive text
		getline(file, line);
		auto pt_count = stoi(line);
		positive_text.clear();
		for (auto i = 0; i < pt_count; ++i)
		{
			getline(file, line);
			positive_text.push_back(line);
		}

		// negative words
		getline(file, line);
		auto nw_count = stoi(line);
		negative_words.clear();
		for (auto i = 0; i < nw_count; ++i)
		{
			getline(file, line);
			negative_words.push_back(line);
		}

		// positive words
		getline(file, line);
		auto pw_count = stoi(line);
		positive_words.clear();
		for (auto i = 0; i < pw_count; ++i)
		{
			getline(file, line);
			positive_words.push_back(line);
		}

		// of << Dn << endl << Dp << endl << D << endl << V << endl << Ln << endl << Lp << endl;
		getline(file, line);
		Dn = stof(line);

		getline(file, line);
		Dp = stof(line);

		getline(file, line);
		D = stof(line);

		getline(file, line);
		V = stof(line);

		getline(file, line);
		Ln = stof(line);

		getline(file, line);
		Lp = stof(line);

		_trained = true;
	}
	catch (...)
	{
		wcerr << L"Can't read " << file_name << endl;
	}

	file.close();

	wcout << L"Model loaded." << endl;

	return _trained;
}

/// Find prediction for text
Predicted Baes::Predict(const wstring& str)
{
	// c = log(Dc / D) + Q[log((Wic + 1) / (V + Lc))]
	if (!_trained)
	{
		wcerr << L"Need to train or load model first." << endl;
		return { NULL };
	}

	auto start_time = clock();
	
	// Классфицируемые слова
	auto words = Split(str, L' ');

	Stemming(words);
		
	// Сколько раз i-ое слово встречалось в документах класса negative в обучающей выборке
	map<wstring, int> Wn;
	// Сколько раз i-ое слово встречалось в документах класса positive в обучающей выборке
	map<wstring, int> Wp;
	
	for (auto &w : words)
	{
		Wn.insert({ w, 0 });
		Wp.insert({ w, 0 });

		// Поиск слова w классифицируемого предложения в классе negative
		auto negative_w = find(begin(negative_words), end(negative_words), w);
		if (negative_w != negative_words.end())
		{
			// Если слово не найдено, то нужно создать
			auto iter = Wn.find(w);
			if (iter == Wn.end())
				Wn.insert({ w, 1 });
			else
				iter->second += 1;
		}

		// Поиск слова w классифицируемого предложения в классе positive
		auto positive_w = find(begin(positive_words), end(positive_words), w);
		if (positive_w != positive_words.end())
		{
			// Если слово не найдено, то нужно создать
			auto iter = Wp.find(w);
			if (iter == Wp.end())
				Wp.insert({ w, 1 });
			else
				iter->second += 1;
		}
	}
	wcout << L"Calculated Wn and Wp: " << double(clock() - start_time) / CLOCKS_PER_SEC << L" sec." << endl;

	// Множество слов классифицируемого документа(включая повторы)
	auto Q = words.size();
	
	double lwp = 0;
	double lwn = 0;
	for (auto &w : words)
	{
		lwp += log(double(Wp[w] + 1) / double(V + Lp));
		lwn += log(double(Wn[w] + 1) / double(V + Ln));
	}

	double Cp = log(double(Dp) / double(D)) + lwp;
	double Cn = log(double(Dn) / double(D)) + lwn;

	double Pp = 1 / double(1 + pow(M_E, Cn - Cp));
	double Pn = 1 - Pp;
		
	return { Pp, Pn, str };
}

Baes::~Baes()
{
#ifdef DEBUG
	wcout << L"~Classificator";
#endif
}

File: baes.py
# coding: utf-8

import re
from nltk.stem import SnowballStemmer
import sys  
import time
import math
import pickle
import argparse

reload(sys)  
sys.setdefaultencoding('utf8')

class Classificator(object):
	""" Baes classificator with count """
	def __init__(self, negative_path, positive_path):
		super(Classificator, self).__init__()
		self.negative_path = negative_path
		self.positive_path = positive_path

		self.negative_text = list()
		self.positive_text = list()

		self.words = list()

	def get_texts(self, count):
		''' Get text from files '''

		# reading nagative
		with open(self.negative_path, "r") as f:
			for i,row in enumerate(f): 
				if i == count:
					break
				try:
					self.negative_text.append(row.split(";")[3][1:-1])
				except Exception, e:
					pass			

		# reading positive
		with open(self.positive_path, "r") as f:
			for i,row in enumerate(f): 
				if i == count:
					break
				try:
					self.positive_text.append(row.split(";")[3][1:-1])
				except Exception, e:
					pass		

	def get_words(self, text_array):
		''' Get words from text array 
		@return array of word in all texts '''

		words = list()
		for t in text_array:
			words.extend(t.split())

		return words

	def filter_words(self, words):
		''' Removing smiles, symbols, hyperlinks '''

		pass

	def stemming(self, words):
		''' Make stem for each word in array 
		@return array of stemming words	'''

		russian_stemmer = SnowballStemmer('russian')
		stemming = list()
		for w in words:
			try:
				stemming.append(russian_stemmer.stem(w))
			except Exception, e:
				pass

		return stemming

	def train(self, count):
		''' Train model
		@return TrainingModel '''

		start_time = time.time()
		self.get_texts(count)
		print "Reading texts: " + str(time.time() - start_time)

		start_time = time.time()
		negative_words = self.stemming(self.get_words(self.negative_text))
		print "Stemming negative words: " + str(time.time() - start_time)

		start_time = time.time()
		positive_words = self.stemming(self.get_words(self.positive_text))
		print "Stemming positive words: " + str(time.time() - start_time)

		return TrainingModel(self.negative_text, self.positive_text, negative_words, positive_words)

	def load(self, model_path):
		with open(model_path, "r") as f:
			return pickle.load(f)

class TrainingModel(object):
	"""Training model"""
	def __init__(self, negative_text, positive_text, negative_words, positive_words):
		super(TrainingModel, self).__init__()
		# Количество текстов в обучающей выборке в классе negative
		self.Dn = len(negative_text)
		# Количество текстов в обучающей выборке в классе positive
		self.Dp = len(positive_text)
		# Общее число текстов в обучающей выборке
		self.D = self.Dn+self.Dp
		# Количество уникальных слов во всех текстах
		# всех уникальных?
		self.V = len(set(negative_words)) + len(set(positive_words))
		# Суммарное количество слов в текстах класса negative
		self.Ln = len(negative_words)
		# Суммарное количество слов в текстах класса positive
		self.Lp = len(positive_words)


		self.positive_words = positive_words
		self.positive_text = positive_text

		self.negative_text = negative_text
		self.negative_words = negative_words

	def save(self, file_name):
		with open(file_name, "w") as f:
			pickle.dump(self, f)
			print "TrainingModel saved to: " + file_name

	def predict(self, text):
		# Сколько раз i-ое слово встречалось в документах класса negative в обучающей выборке
		#Wn
		# Сколько раз i-ое слово встречалось в документах класса positive в обучающей выборке
		#Wp
		# Множество слов классифицируемого документа (включая повторы)
		#Q

		# разбор текста на слова и прогоняем через стеммер
		russian_stemmer = SnowballStemmer('russian')
		words = list()
		for w in text.split():
			try:
				words.append(russian_stemmer.stem(w))
			except:
				pass

		Wp = dict.fromkeys(words, 0)
		Wn = dict.fromkeys(words, 0)

		for w in words:
			if w in self.positive_words:
				Wp[w] += 1
			if w in self.negative_words:
				Wn[w] += 1

		Q = len(words)

		#c = log(Dc/D) + Q[log((Wic + 1)/(V+Lc))]
		lwp = 0
		lwn = 0
		for i,w in enumerate(words):
			lwp += math.log( float(Wp[w]+1) / float(self.V+self.Lp) )
			lwn += math.log( float(Wn[w]+1) / float(self.V+self.Ln) )

		Cp = math.log(float(self.Dp) / float(self.D)) + lwp
		Cn = math.log(float(self.Dn) / float(self.D)) + lwn

		Pp = 1 / float(1 + math.e ** (Cn - Cp) ) 
		Pn = 1 - Pp

		return {'negative': Pn, 'positive': Pp, 'text': text}

if __name__ == "__main__":
	parser = argparse.ArgumentParser(description='Train and predict.')
	subparsers = parser.add_subparsers(help='List of commands')

	train_parser = subparsers.add_parser('train', help='Training model')
	train_parser.add_argument('--negative', help='Negative.csv path', required=True)
	train_parser.add_argument('--positive', help='Positive.csv path', required=True)
	train_parser.add_argument('--model', help='Model path to save', required=True)
	train_parser.add_argument('--count', help='Training count', type=int, required=True)
	train_parser.set_defaults(which='train')

	predict_parser = subparsers.add_parser('predict', help='Predict model')
	predict_parser.add_argument('--model', required=True, help='Model path to load')
	predict_parser.set_defaults(which='predict')

	args = parser.parse_args()
	if args.which == 'train':
		start_time = time.time()

		baes = Classificator(args.negative, args.positive)
		model = baes.train(args.count)
		model.save(args.model)

		print "Execution time: " + str(time.time() - start_time)
	elif args.which == 'predict':
		baes = Classificator(None, None)
		start_time = time.time()
		model = baes.load(args.model)
		print "Loading: " + str(time.time() - start_time)

		test = list()
		test.append(model.predict("Пост хрень, да и всё херово кончилось в фильме :("))
		test.append(model.predict("Нарвалы любят единорогов больше жизни"))
		test.append(model.predict("Корабли лавировали лавировали, да всё хорошо кончилось :)"))
		test.append(model.predict("Меня на выходе дверью пришибло, поэтому на троечку"))
		test.append(model.predict("Хамло!"))
		test.append(model.predict("Да успокойтесь, барышня, всё не так плохо ;-)"))

		for x in test:
			print x.get("text") + ": positive=" + str(x.get("positive")) + ", negative=" + str(x.get("negative"))

		print "Execution time: " + str(time.time() - start_time)
